source/extensions/filters/http/aws_lambda/config.cc

```diff
@@ -1,21 +1,139 @@
 #include "extensions/filters/http/aws_lambda/config.h"
 
+#include "envoy/thread_local/thread_local.h"
+
 namespace Envoy {
 namespace Extensions {
 namespace HttpFilters {
 namespace AwsLambda {
 
+namespace CommonAws = Envoy::Extensions::HttpFilters::Common::Aws;
+
+namespace {
+
+// Current AWS implementation will only refresh creds if *more* than an hour has
+// passed. In the AWS impl these creds are fetched on the main thread. As we use
+// the creds in the filter, we cache the creds in a TLS slot, so they are
+// available to the filters in the worker threads. As we have our own cache, we
+// have to implement our own refresh loop. To avoid being to tied down the
+// timing of the AWS implementation, we'll have a refresh interval of a few
+// minutes. refresh itself will only happen when deemed needed by the AWS code.
+//
+// According to the AWS docs
+// (https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_use_switch-role-ec2.html)
+//
+// >>> it should get a refreshed set of credentials every hour, or at least 15
+// >>> minutes before the current set expires.
+//
+// Refreshing every 14 minutes should guarantee us fresh credentials.
+constexpr std::chrono::milliseconds REFRESH_AWS_CREDS =
+    std::chrono::minutes(14);
+
+struct ThreadLocalState : public Envoy::ThreadLocal::ThreadLocalObject {
+  ThreadLocalState(CredentialsConstSharedPtr credentials)
+      : credentials_(credentials) {}
+  CredentialsConstSharedPtr credentials_;
+};
+
+} // namespace
+
+AWSLambdaConfigImpl::AWSLambdaConfigImpl(
+    std::unique_ptr<Common::Aws::CredentialsProvider> &&provider,
+    Event::Dispatcher &dispatcher, Envoy::ThreadLocal::SlotAllocator &tls,
+    const envoy::config::filter::http::aws_lambda::v2::AWSLambdaConfig
+        &protoconfig) {
+  bool use_default_credentials = false;
+
+  if (protoconfig.has_use_default_credentials()) {
+    use_default_credentials = protoconfig.use_default_credentials().value();
+  }
+
+  if (use_default_credentials) {
+    provider_ = std::move(provider);
+
+    tls_slot_ = tls.allocateSlot();
+    auto empty_creds = std::make_shared<const CommonAws::Credentials>();
+    tls_slot_->set([empty_creds](Event::Dispatcher &) {
+      return std::make_shared<ThreadLocalState>(empty_creds);
+    });
+
+    timer_ = dispatcher.createTimer([this] { timerCallback(); });
+    // call the time callback to fetch credentials now.
+    // this will also re-trigger the timer.
+    timerCallback();
+  }
+}
+
+CredentialsConstSharedPtr AWSLambdaConfigImpl::getCredentials() const {
+  if (!provider_) {
+    return {};
+  }
+
+  // tls_slot_ != nil IFF provider_ != nil
+  return tls_slot_->getTyped<ThreadLocalState>().credentials_;
+}
+
+void AWSLambdaConfigImpl::timerCallback() {
+  // get new credentials.
+  auto new_creds = provider_->getCredentials();
+  if (new_creds == CommonAws::Credentials()) {
+    ENVOY_LOG(warn, "can't get AWS credentials - credentials will not be "
```

Comment:
No stream log? is that because this is in the filter config, and therefore out of the request thread?

Reply:
correct; the config processing happens in the main thread. there is no stream in this context.